fish-rust/src/event.rs

```diff
@@ -0,0 +1,926 @@
+//! Functions for handling event triggers
+//!
+//! Because most of these functions can be called by signal handler, it is important to make it well
+//! defined when these functions produce output or perform memory allocations, since such functions
+//! may not be safely called by signal handlers.
+
+use autocxx::WithinUniquePtr;
+use cxx::{CxxVector, CxxWString, UniquePtr};
+use libc::pid_t;
+use std::pin::Pin;
+use std::sync::atomic::{AtomicBool, AtomicU32, Ordering};
+use std::sync::{Arc, Mutex};
+use widestring_suffix::widestrs;
+
+use crate::bitset::BitSet;
+use crate::builtins::shared::io_streams_t;
+use crate::common::{escape_string, EscapeFlags, EscapeStringStyle};
+use crate::ffi::{
+    self, block_t, parser_t, signal_check_cancel, signal_handle, termsize_container_t, Repin,
+};
+use crate::flog::FLOG;
+use crate::signal::{sig2wcs, signal_get_desc};
+use crate::wchar::{wstr, WString, L};
+use crate::wchar_ffi::{wcharz_t, AsWstr, WCharFromFFI, WCharToFFI};
+use crate::wutil::sprintf;
+
+#[cxx::bridge]
+mod event_ffi {
+    extern "C++" {
+        include!("wutil.h");
+        include!("parser.h");
+        include!("io.h");
+        type wcharz_t = crate::ffi::wcharz_t;
+        type parser_t = crate::ffi::parser_t;
+        type io_streams_t = crate::ffi::io_streams_t;
+    }
+
+    enum event_type_t {
+        any,
```

Comment:
duplicating the enum seems great because it makes it easy to get the post-FFI cleanup step right